Change the world refactor (#75)
substratusai/substratus: docs/design.md

Optionally, a loader can be specified to load ML models weights and biases from specific sources.
For example, a commonly used model loader would be the HuggingFace model loader
where by providing the model name would load the model to the Model object
bucket location. Model loaders are defined as container images and can either
point to prebuilt container images or a Git repository with Dockerfile.

A model

The Model object can optionally describe the training parameters including a reference to the Dataset that should be used for training. In this case, the Model object represents the eventual output of a training/finetuning operation.

```yaml
kind: Model
spec:
  container: # used for either loader or trainer
  loader: {} # Optional, when empty an just a location will be created with no files in it
  trainer: {} # Optional, will run a training job
```

### Model Sources

A Model can be sourced from Git or a prebuilt image. Using Git will trigger a container build Job.

```yaml
kind: Model
name: falcon-7b
spec:
  container:
    git:
      url: https://github.com/substratusai/hf-model-loader
    image: substratusai/hf-model-loader
  loader:

>     params: {name: tiuu/falcon-7b}

Reviewer: can we convert any param.x to PARAM_X as env variable?

So params.name would be available as env variable `PARAM_NAME` in the container itself.

What you think? @brandonjbjelland @nstogner

Author: https://github.com/substratusai/substratus/pull/75#discussion_r1257490669

    # optional future

>     paramsFrom: configMap?

Reviewer: ```suggestion
  status:
    url: gcs://my-bucket/my-model/
    files: ["pytorch-001.bin", "config.json"]
    size: 27Gi
```

```

`params` get converted to environment variables following a specific scheme

The controller will orchestrate the following flow in this case:

<img src="./diagrams/model-building.excalidraw.png" width="80%"></img>

### Model Training

Approach 1: Include in Model object

Models can be trained by specifying the `.spec.training` section.

This will orchestrate a training Job with multiple FUSE mounted buckets.

<img src="./diagrams/model-training.excalidraw.png" width="80%"></img>

```yaml
kind: Model
name: falcon-7b-k8s
spec:
  container:
  # can only specify git or image not both
    git: 
      url: https://github.com/substratusai/hf-llm-trainer
      ref: refs/branch/main
    image: substratusai/hf-llm-peft-trainer # eventually introduce native pytorch trainer

  trainer:
    # sourceModel can be left out for creating new base model
    sourceModel:
      name: falcon-7b
      namespace: other-namespace
    dataset:
      name: k8s-instructions
      namespace: other-namespace

>     resources: {}

Reviewer: should resources be under `container`?

Author: Currently everything under container is a source of the container... this would break that pattern. I vote `.spec.resources` because that way we can keep it consistent across all object kinds.

    params: {epochs: 1}
```

> 
> 

Reviewer 2: ```suggestion

```

### Model Training with Notebook
Notebook container contract is to run notebook.sh and serve
notebook compatible HTTP endpoint on port 8888.

User flow: Random experiments

> notebook-gpu image has transformers, pytorch, cuda, python 3, etc.. basically most you need

Reviewer 2: ```suggestion
notebook-gpu image has transformers, pytorch, cuda, python 3, etc.. most everything you need to get started.
```

example notebook in notebook.

1. Create Notebook with no sourceModel
2. Load a model dynamically in my notebook session
3. Experiment with the model and training and only save my notebook somewhere for reference

```yaml
kind: Notebook
name: notebooks-for-anything
spec:
  container:
    # or git repo as a source
    image: substratusai/notebook-gpu

>   resources: {gpuCount: 4, type: L4}

Reviewer: I think resources should be exactly the same as container resources:
```
    resources:
      requests:
        memory: "10Gi"
        cpu: "10"
      limits:
       nvidia.com/gpu: 2
```
would be valid here.

The open question then would be how to pass it GPU type. Should that be a separate attribute outside of resources?  I vote for yes gpu_type has separate attribute such that we can use standard K8s resource requests which most are familiar with.

Author: I am not sure we should directly expose kubernetes `.resources`. Mainly because of what you called out here... the use of specialized equipment like GPUs requires a slew of other fields to be set. I think substratus should simplify this by exposing a few high-level resource knobs. Perhaps:

```yaml
runtime:
  compute: ["GPU", "CPU"] # Can be run on either GPUs or CPUs.
  strategy: Fastest # or "Cheapest", or "Balanced"
  memory: "10Gi" # Translates to AT LEAST 10Gi of gpu-memory and 10Gi or cpu-memory. If more is provisioned (commonly the case), it will be reported to the container using environment variables.
  disk: "50Gi"
```

CPU cores are not really a requirement... more will be faster.

Reviewer: When you define a model you have to decide if you want to run the loader on CPU or GPU

Alternative proposal for runtime under loader or trainer section
```
runtime:
  accelerator: # optional if not specified will only use CPU
     type: nvidia-tesla-a100
     count: 8
  cpu: 2 # default will always be 2 if not specified
  memory: "10Gi" # will always be 10 if not specified. Translates to AT LEAST 10Gi of gpu-memory and 10Gi or cpu-memory. If more is provisioned (commonly the case), it will be reported to the container using environment variables.
  disk: "50Gi" # will always be 50Gi if not specified
```

Author: When the specify the accelerator in the Model object, the Model object is no longer portable across clusters that have access to different accelerators.

Reviewer: I want to specify exactly the resources for my training job. The model loaded in GCS is still portable across clusters. I think I'm misunderstanding the portability concern

Reviewer 2: Perhaps there's a middleground where `runtimeProfile`s can be created by administrators and used by consumers in the form you're pitching @nstogner . Personally, I lean toward starting with the more raw resource spec model since model requirements are quite specific.

Reviewer 2: > I want to specify exactly the resources for my training job. The model loaded in GCS is still portable across clusters. I think I'm misunderstanding the portability concern

Guessing at what Nick meant here: where might models not achieve portability:
1. `resource_limits` might differ from cluster to cluster.
2. different accelerator types are supported in different regions/zones and across different cloud providers.
3. even in the case where both of the above criteria are met, stock out might mean another accelerator type must be used if the job is to proceed.

I think at the moment I still lean toward raw resource specs as @samos123 is pushing for here but potentially a way to think about this is giving an ordered list of hardware specs so if one resource profile fails, another one can be tried.

Author: Resolved in slack

```

> User flow: Iterate on model training code

Reviewer 2: I think this captures the use case a bit more precisely:

```suggestion
Lab user flow: Iterate on model training code using a customized trainer image
```

> 1. `git clone https://substratusai/hf-llm-trainer && cd hf-llm-trainer`
> 2. Modify Dockerfile
> 2. Create a Notebook that mounts source model
> 3. `kubectl open notebook notebook-training-experiment`
>    * tar up directory respecting dockerignore
>    * Get GCS signed url from controller notebook status
>    * upload to workspace.tar to signed URL
> 3. Notebook automatically has all my local directory in the notebook when using the kubectl plugin

Reviewer 2: This probably looks different in the VSCode extension-based flow but using `jupyter lab`:

```suggestion
1. `git clone https://substratusai/hf-llm-trainer && cd hf-llm-trainer`
2. Modify Dockerfile
3. Create a Notebook manifest and apply the Notebook CR mounting a source model
4. `kubectl open notebook -f notebook-training-experiment` - the `kubectl` plugin does each of the following additional steps when `spec.container.local = true`
   * tar up directory respecting dockerignore
   * Get GCS signed url from controller notebook status 
   * upload to workspace.tar to signed URL
5. Notebook automatically has the local directory contents in the notebook when using the `kubectl` plugin (via cp sync operations)
```

```yaml
kind: Notebook
name: notebook-training-experiment
spec:
  container:
    local: true
  sourceModel: falcon-7b
  resources: {}
status:
    uploadUrl: gs://substratusai-notebooks/workspace.tar
```

> How to get from Notebook to trainer?
> 
> Sync back files from the notebook into local directory
> and built into new image.

Reviewer 2: This is an unfinished thought as we somewhat ran out of steam but here's my take:

```suggestion
User flow: How to get from Notebook to trainer?

1. As `kubectl open notebook ...` is terminated by the user, files will be synced from the Notebook back to the local directory. We can hope but not guarantee this is a git repo.
2. Optional: During `kubectl open notebook ...` termination, a signal will be sent to the controller to build a container image which can serve as the trainer.
3. Potentially the user is prompted for this auto-generation: A `model-${epoch}.yaml` manifest having the values of that trainer will be generated into the repo root having either spec.container.image populated with the build from step 2 OR the repo/branch info taken from the local dir. 

  ```yaml
  kind: Model
  name: falcon-7b-k8s-${epoch_or_branch_name}
  spec:
    container:
      image: my-org/model-falcon-7b-k8s-trainer # the just-built trainer image
  
    trainer:
      sourceModel:
        name: falcon-7b
        namespace: other-namespace
      dataset:
        name: k8s-instructions
        namespace: other-namespace
      resources: {}
      params:
        epochs: 1
  ```

4. The user could be provided a command to submit their training job.

### Model Serving
```yaml
kind: ModelServer
spec:
  container:
    image: substratusai/basaran # or git repo
  # model gets mounted to /model/saved
  model: falcon-7b-k8s
  # optional
  quantization:
    bits: 4 # default is 16
  resources: {}

> ```

Reviewer 2: I like this in particular! 👏

# Design

## Models

A Model object represents a logical ML model.

The object describes where the source code for the model should come from. This code will be stored in a container image. That container image will initially be run in order to store the initial weights and biases in a bucket that is mounted into the container as a directory. This is most useful in cases where the Model is representing a pre-trained model.

The Model object can optionally describe the training parameters including a reference to the Dataset that should be used for training. In this case, the Model object represents the eventual output of a training/finetuning operation.

```yaml
kind: Model
spec:

>   source: {} # Required

Reviewer: could we make source optional such that you can create a notebook with an empty model attached?

Reviewer: That would also allow us to have a FinetuningJob CRD later if needed

  training: {} # Optional, will run a training job
```

### Model Sources

A Model can be sourced from Git. This will trigger a build Job.

```yaml
kind: Model
spec:
  source:
    git:

>       url: https://github.com/substratusai/model-falcon-7b

Reviewer: I guess in this case it could be 
```
url: https://github.com/substratusai/model-loader-huggingface
name: falcon-7b
```

The huggingface model loader can load any model from huggingface into the GCS bucket
